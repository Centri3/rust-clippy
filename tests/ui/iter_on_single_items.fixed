//@run-rustfix
#![warn(clippy::iter_on_single_items)]
#![allow(clippy::iter_next_slice, clippy::redundant_clone)]

fn array() {
    assert_eq!(std::iter::once(123).next(), Some(123));
    assert_eq!(std::iter::once(&mut 123).next(), Some(&mut 123));
    assert_eq!(std::iter::once(&123).next(), Some(&123));
    assert_eq!(std::iter::once(123).next(), Some(123));
    assert_eq!(std::iter::once(&mut 123).next(), Some(&mut 123));
    assert_eq!(std::iter::once(&123).next(), Some(&123));

    let x: [u32; 1] = [1];
    x.iter().copied().chain(std::iter::once(1));
    x.iter().copied().eq(std::iter::once(1));
    x.iter().copied().ne(std::iter::once(1));
    let x: Vec<u32> = vec![1];
    x.iter().copied().chain(std::iter::once(1));
    x.iter().copied().eq(std::iter::once(1));
    x.iter().copied().ne(std::iter::once(1));
    let x: Vec<u32> = vec![];
    x.iter().copied().chain(std::iter::once(0));
    x.iter().copied().eq(std::iter::once(0));
    x.iter().copied().ne(std::iter::once(0));

    // Don't trigger on non-iter methods
    let _: Option<String> = Some("test".to_string()).clone();
    let _: [String; 1] = ["test".to_string()].clone();

    // Don't trigger on match or if branches
    let _ = match 123 {
        123 => [].iter(),
        _ => ["test"].iter(),
    };

    let _ = if false { ["test"].iter() } else { [].iter() };
}

macro_rules! in_macros {
    () => {
        assert_eq!(std::iter::once(123).next(), Some(123));
        assert_eq!(std::iter::once(&mut 123).next(), Some(&mut 123));
        assert_eq!(std::iter::once(&123).next(), Some(&123));
        assert_eq!(std::iter::once(123).next(), Some(123));
        assert_eq!(std::iter::once(&mut 123).next(), Some(&mut 123));
        assert_eq!(std::iter::once(&123).next(), Some(&123));
    };
}

// Don't trigger on a `Some` that isn't std's option
mod custom_option {
    #[allow(unused)]
    enum CustomOption {
        Some(i32),
        None,
    }

    impl CustomOption {
        fn iter(&self) {}
        fn iter_mut(&mut self) {}
        fn into_iter(self) {}
    }
    use CustomOption::*;

    pub fn custom_option() {
        Some(3).iter();
        Some(3).iter_mut();
        Some(3).into_iter();
    }
}

fn main() {
    array();
    custom_option::custom_option();
    in_macros!();
}
