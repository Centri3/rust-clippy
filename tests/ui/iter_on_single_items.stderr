error: usage of `[T; 1]` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:6:16
   |
LL |     assert_eq!([123].into_iter().next(), Some(123));
   |                ^^^^^------------
   |                |
   |                help: use `std::iter::once` instead: `std::iter::once(123)`
   |
   = note: `-D clippy::iter-on-single-items` implied by `-D warnings`

error: usage of `[T; 1]` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:7:16
   |
LL |     assert_eq!([123].iter_mut().next(), Some(&mut 123));
   |                ^^^^^-----------
   |                |
   |                help: use `std::iter::once` instead: `std::iter::once(&mut 123)`

error: usage of `[T; 1]` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:8:16
   |
LL |     assert_eq!([123].iter().next(), Some(&123));
   |                ^^^^^-------
   |                |
   |                help: use `std::iter::once` instead: `std::iter::once(&123)`

error: usage of `Some` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:9:16
   |
LL |     assert_eq!(Some(123).into_iter().next(), Some(123));
   |                ^^^^^^^^^------------
   |                |
   |                help: use `std::iter::once` instead: `std::iter::once(123)`

error: usage of `Some` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:10:16
   |
LL |     assert_eq!(Some(123).iter_mut().next(), Some(&mut 123));
   |                ^^^^^^^^^-----------
   |                |
   |                help: use `std::iter::once` instead: `std::iter::once(&mut 123)`

error: usage of `Some` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:11:16
   |
LL |     assert_eq!(Some(123).iter().next(), Some(&123));
   |                ^^^^^^^^^-------
   |                |
   |                help: use `std::iter::once` instead: `std::iter::once(&123)`

error: usage of `[T; 1]` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:14:29
   |
LL |     x.iter().copied().chain([1]);
   |                             ^^^ help: use `std::iter::once` instead: `std::iter::once(1)`
   |
   = note: this method is generic over `Iterator`

error: usage of `[T; 1]` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:15:26
   |
LL |     x.iter().copied().eq([1]);
   |                          ^^^ help: use `std::iter::once` instead: `std::iter::once(1)`
   |
   = note: this method is generic over `Iterator`

error: usage of `[T; 1]` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:16:26
   |
LL |     x.iter().copied().ne([1]);
   |                          ^^^ help: use `std::iter::once` instead: `std::iter::once(1)`
   |
   = note: this method is generic over `Iterator`

error: usage of `[T; 1]` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:18:29
   |
LL |     x.iter().copied().chain(vec![1]);
   |                             ^^^^^^^ help: use `std::iter::once` instead: `std::iter::once(1)`
   |
   = note: this method is generic over `Iterator`

error: usage of `[T; 1]` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:19:26
   |
LL |     x.iter().copied().eq(vec![1]);
   |                          ^^^^^^^ help: use `std::iter::once` instead: `std::iter::once(1)`
   |
   = note: this method is generic over `Iterator`

error: usage of `[T; 1]` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:20:26
   |
LL |     x.iter().copied().ne(vec![1]);
   |                          ^^^^^^^ help: use `std::iter::once` instead: `std::iter::once(1)`
   |
   = note: this method is generic over `Iterator`

error: usage of `[T; 1]` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:22:29
   |
LL |     x.iter().copied().chain(vec![0; 1]);
   |                             ^^^^^^^^^^ help: use `std::iter::once` instead: `std::iter::once(0)`
   |
   = note: this method is generic over `Iterator`

error: usage of `[T; 1]` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:23:26
   |
LL |     x.iter().copied().eq(vec![0; 1]);
   |                          ^^^^^^^^^^ help: use `std::iter::once` instead: `std::iter::once(0)`
   |
   = note: this method is generic over `Iterator`

error: usage of `[T; 1]` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:24:26
   |
LL |     x.iter().copied().ne(vec![0; 1]);
   |                          ^^^^^^^^^^ help: use `std::iter::once` instead: `std::iter::once(0)`
   |
   = note: this method is generic over `Iterator`

error: usage of `[T; 1]` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:75:5
   |
LL |         assert_eq!([123].into_iter().next(), Some(123));
   |                    ----------------- help: use `std::iter::once` instead: `std::iter::once(123)`
...
LL |     in_macros!();
   |     ^^^^^^^^^^^^

error: usage of `[T; 1]` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:75:5
   |
LL |         assert_eq!([123].iter_mut().next(), Some(&mut 123));
   |                    ---------------- help: use `std::iter::once` instead: `std::iter::once(&mut 123)`
...
LL |     in_macros!();
   |     ^^^^^^^^^^^^

error: usage of `[T; 1]` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:75:5
   |
LL |         assert_eq!([123].iter().next(), Some(&123));
   |                    ------------ help: use `std::iter::once` instead: `std::iter::once(&123)`
...
LL |     in_macros!();
   |     ^^^^^^^^^^^^

error: usage of `Some` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:75:5
   |
LL |         assert_eq!(Some(123).into_iter().next(), Some(123));
   |                    --------------------- help: use `std::iter::once` instead: `std::iter::once(123)`
...
LL |     in_macros!();
   |     ^^^^^^^^^^^^

error: usage of `Some` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:75:5
   |
LL |         assert_eq!(Some(123).iter_mut().next(), Some(&mut 123));
   |                    -------------------- help: use `std::iter::once` instead: `std::iter::once(&mut 123)`
...
LL |     in_macros!();
   |     ^^^^^^^^^^^^

error: usage of `Some` to create an iterator with only one element
  --> $DIR/iter_on_single_items.rs:75:5
   |
LL |         assert_eq!(Some(123).iter().next(), Some(&123));
   |                    ---------------- help: use `std::iter::once` instead: `std::iter::once(&123)`
...
LL |     in_macros!();
   |     ^^^^^^^^^^^^

error: aborting due to 21 previous errors

